### PR TITLE
fix(applicationset): prevent applicationset progressive sync from stalling in pending

### LINKED ISSUE #12459: Applicationset progressive sync does not proceed
Checklist:

* [x] I've searched in the docs and FAQ for my answer: https://bit.ly/argocd-faq.
* [x] I've included steps to reproduce the bug.
* [x] I've pasted the output of `argocd version`.

**Describe the bug**

I have an applicationset using a git generator and the following rolling sync strategy to update one app (the `prerelease`) before the others:
```
- matchExpressions:
    - key: service
      operator: In
      values:
        - prerelease
- matchExpressions:
    - key: service
      operator: NotIn
      values:
        - prerelease
```
The application with the `service:prerelease` label is successfully triggered. Runs its predeploy hook, and the deployment gets updated. Once complete, the prerelease app is healthy, but the remaining apps are not triggered to sync, and sit at "out of sync" indefinitely. Manually triggering a sync action again on the prerelease app re-runs the sync operation, which then successfully triggers the other apps to sync.

I can see the logs in the applicationset controller indicating the expected apps are in the proper steps, but `Application allowed to sync before maxUpdate` for the step2 apps continues to show false even after the step1 app has is healthy

**To Reproduce**

Create an applicationset that uses a git generator and a rolling sync strategy that requires a single application to update before the others

**Expected behavior**

All apps in subsequent sync steps are triggered.

**Version**
v2.6.1+3f143c9

**Logs**
app & appset names truncated/redacted
```
time="2023-02-14T18:17:35Z" level=info msg="Application prerelease has entered Progressing status, updating its ApplicationSet status to Progressing"
time="2023-02-14T18:17:35Z" level=info msg="ApplicationSet <APPSET_NAME> step list:"
time="2023-02-14T18:17:35Z" level=info msg="step 1: [prerelease]"
time="2023-02-14T18:17:35Z" level=info msg="step 2: [app, analytics, metrics]"
time="2023-02-14T18:17:35Z" level=info msg="Application allowed to sync before maxUpdate?: map[app:false, analytics:false, metrics:false]"
time="2023-02-14T18:17:35Z" level=info msg="unchanged Application" app=analytics appSet=<APPSET_NAME>
time="2023-02-14T18:17:35Z" level=info msg="unchanged Application" app=app appSet=<APPSET_NAME>
time="2023-02-14T18:17:35Z" level=info msg="unchanged Application" app=metrics appSet=l<APPSET_NAME>
time="2023-02-14T18:17:35Z" level=info msg="end reconcile" applicationset=argocd/<APPSET_NAME> requeueAfter=3m0s
```
The `allowed to sync` remains false on the subsequent reconciliation

### PR DESCRIPTION
**Pull request by wparr-circle**
Fixes https://github.com/argoproj/argo-cd/issues/12459

ApplicationSet Controller Progressive Sync sometimes looks to get into a state where the Applications are never synced and a rollout never concludes, based on the timestamps where an ApplicationSet may have transitioned as time T and triggered a sync - but somehow an Application Sync Operation indicates that it was triggered much earlier at T-10..N.
The disparity looks to possibly be in the order of minutes, we've seen it happen with ~20-30 seconds of disparity. However based on #12459 it looks like could be much more.
There is some existing behaviour implemented in https://github.com/argoproj/argo-cd/pull/13926 - which manages to resolve cases where the disparity is lesser than 10s - but doesn't help resolve cases where it exceed this.

This pull resolves the issue by setting an identifier for a sync triggered by the ApplicationSet controller, so it may be compared to the details of an operation ie. it was triggered by the applicationset controller and has the expected id, while still allowing other syncs triggered by other users to result in transitioning from Pending to Progressing provided the sync occured after ApplicationSet Controller transitioned the Application to Pending.

This does not patch a regression, as it's affected previous versions as early as 2.6, 2.7 and still affects 2.10. It may require cherry picking to older versions.

Checklist:

* [X] Either (a) I've created an [enhancement proposal](https://github.com/argoproj/argo-cd/issues/new/choose) and discussed it with the community, (b) this is a bug fix, or (c) this does not need to be in the release notes.
* [x] The title of the PR states what changed and the related issues number (used for the release note).
* [x] The title of the PR conforms to the [Toolchain Guide](https://argo-cd.readthedocs.io/en/latest/developer-guide/toolchain-guide/#title-of-the-pr)
* [x] I've included "Closes [ISSUE #]" or "Fixes [ISSUE #]" in the description to automatically close the associated issue.
* [x] I've updated both the CLI and UI to expose my feature, or I plan to submit a second PR with them.
* [ ] Does this PR require documentation updates?
* [ ] I've updated documentation as required by this PR.
* [x] I have signed off all my commits as required by [DCO](https://github.com/argoproj/argoproj/blob/master/community/CONTRIBUTING.md#legal)
* [x] I have written unit and/or e2e tests for my change. PRs without these are unlikely to be merged.
* [x] My build is green ([troubleshooting builds](https://argo-cd.readthedocs.io/en/latest/developer-guide/ci/)).
* [ ] My new feature complies with the [feature status](https://github.com/argoproj/argoproj/blob/master/community/feature-status.md) guidelines.
* [x] I have added a brief description of why this PR is necessary and/or what this PR solves.
* [ ] Optional. My organization is added to USERS.md.
* [x] Optional. For bug fixes, I've indicated what older releases this fix should be cherry-picked into (this may or may not happen depending on risk/complexity).

### CODE REVIEW

**Review by wparr-circle — COMMENTED** [review]
Probably a missing generation of a fresh ID for an out dated application. Will follow up in the morning.

**Comment by jannfis** [comment]
Would this fix be possible without a CRD change? Usually, we have a policy of not cherry-picking changes to CRDs back into already released branches, for update simplicity.

If it's not possible without CRD change, I think we may want to forward-fix this only.

**Comment by jannfis** [comment]
@wparr-circle I took a look at the PR in your fork, and I also like that change better than the original. It's much simpler and easier to read.

> However, I'm not 100% if these fields are always available as I am assuming.

You will get the target revisions as long as the application has synced once, otherwise, you'll get an empty list. What exactly is your assumption about this field, i.e. what properties does it need to match to suite your logic?

**Comment by wparr-circle** [comment]
My assumption was exactly this:
```
You will get the target revisions as long as the application has synced at least once, otherwise, you'll get an empty list
```

And that the fields of the struct:

- `app.Status.OperationState.SyncResult.Revisions` / `app.Status.OperationState.SyncResult.Revision` - Corresponds to the last revision that a sync operation resolved to.
- `app.Sync.Revisions` - Corresponds to the last revision that a reconcile resolved to. ie. when out of sync this will be different from the former when auto-sync is disabled or the sync operation otherwise hasn't run yet.

**Comment by wparr-circle** [comment]
Following up on this @jannfis, any for a maintainer to have a look over this.

**Comment by crenshaw-dev** [comment]
@wmgroot any concerns about the approach?

File: applicationset/controllers/applicationset_controller.go [hunk 1]
```diff
@@ -1169,20 +1170,27 @@ func (r *ApplicationSetReconciler) updateApplicationSetApplicationStatus(ctx con
 			currentAppStatus.Status = "Waiting"
 			currentAppStatus.Message = "Application has pending changes, setting status to Waiting."
 			currentAppStatus.Step = fmt.Sprint(appStepMap[currentAppStatus.Application] + 1)
+			currentAppStatus.TargetRevisions = app.Status.GetRevisions()
 		}
 
 		if currentAppStatus.Status == "Pending" {
 			// check for successful syncs started less than 10s before the Application transitioned to Pending
 			// this covers race conditions where syncs initiated by RollingSync miraculously have a sync time before the transition to Pending state occurred (could be a few seconds)
```
Comment (wmgroot) [inline applicationset/controllers/applicationset_controller.go:1178]: Lets update or remove these comments to prevent confusion in the future.

**Comment by wmgroot** [comment]
This approach looks good to me.

**Comment by fcrespofastly** [comment]
@wparr-circle just tagging you just in case the previous message from @wmgroot came unnoticed. I think the PR is good to go! good job!!

File: applicationset/controllers/applicationset_controller.go (hunk 1 above)
Comment (wparr-circle) [reply applicationset/controllers/applicationset_controller.go]: Updated this comment!

**Comment by fcrespofastly** [comment]
@wmgroot can we merge this?

**Comment by wmgroot** [comment]
Sure, go for it.

**Comment by fcrespofastly** [comment]
> Sure, go for it.
I wish I could! 😅

**Comment by crenshaw-dev** [comment]
lol let me run a quick test locally, then I'll merge. I trust Matt's judgment but also it's always good to get one local run in. :-)

**Comment by aliazlan-t** [comment]
@crenshaw-dev can we merge this now?

**Comment by gmauleon** [comment]
Hello good sirs, does this have a chance to make it into 2.11? 😁

**Comment by fcrespofastly** [comment]
Hey @crenshaw-dev 👋🏻 !

This PR potentially fixes a few issues re stale syncs, do you feel more comfortable merging after testing it? 

Thank you!!

**Comment by wparr-circle** [comment]
Following up here @crenshaw-dev, did you get a chance to run locally? Any concerns? Thanks

**Comment by fcrespofastly** [comment]
Hey @wmgroot @crenshaw-dev 👋🏻  

What can we do to help you get this merged?

Thanks!!

**Comment by offzale** [comment]
@wparr-circle thanks a lot for this fix! I have been facing this issue ever since I started using progressive syncs last year.
@crenshaw-dev any chance to get this merged soon?

**Comment by wparr-circle** [comment]
Plus one on the above @crenshaw-dev. It would be great to get this merged off, as we would like to be able to use the feature 🙏

Anything I can do to help with any concerns, or just not had time?

**Review by crenshaw-dev — APPROVED** [review]

**Comment by offzale** [comment]
After a week from applying this fix and hundreds of syncs being triggered, the issue seems to be gone. Thanks a lot once again @wparr-circle! You saved me a lot of time on watching and triggering manunal syncs. 🙏 

What I have noticed though is that the last application in the appset is left sometimes in a `Pending` status after being synced. This happens some times, and it does not prevent the appset from triggering the next sync. After a new sync that status on the application is back to Healthy, some times as well. So, it does not affect negatively the appset sync capabilities, I just saw it as I was watching the appsets' applications closely to monitor the fix.

Anybody else noticed this?

**Comment by wparr-circle** [comment]
Thanks for sharing @offzale, glad you're not seeing the same issue any more!

I haven't noticed any cases where the _last_ application in the AppSet is left in a Pending status personally. I'll keep an eye out for anything like this as we roll out the new version 👀 
 
Is there anything specific that seems to an issue with the Application? 
Could you share status, events or logs from around the time you observe it happening?

**Pull request merged**